### PR TITLE
Split node into one file per module

### PR DESCRIPTION
**Pull request by rbuckton**
Per https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30477#issuecomment-445801787 and https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30477#issuecomment-445835707, this splits node's definitions into one file per module in advance of leveraging `"typesVersions"` support.

### CODE REVIEW

**Comment by rbuckton** [comment]
@RyanCavanaugh, @DanielRosenwasser: Do you have any thoughts on this change? This was recommended by @Flarna and @KonanMentor as something worth doing as part of adding `"typesVersions"` support to node.

File: types/node/process.d.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+declare module "process" {
+    export = process;
```
Comment (SimonSchick) [inline types/node/process.d.ts]: Nit: Shouldn't this file contain the actual process definitions and the global should reference this?

File: types/node/module.d.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+declare module "module" {
+    export = NodeJS.Module;
```
Comment (SimonSchick) [inline types/node/module.d.ts]: See comment at module.d.ts

**Comment by Flarna** [comment]
Honestly speaking I'm still not sure if typesversion is something we want to have at all as it just increases the maintenance burden - it's still needed to type everything for 2.1. Personally I would prefer to raise the minimum typescript version like it was done in other modules (e.g. react used min version 2.8). Are there any plans to move forward in this?

Currently the use if it blocks node changes anyway as CI fails in jsdom and since a few hours additionally in google-cloud__tasks (see #32444).

If the decisions is to use typesversions splitting files is a good idea in my opinion. Most likely tests need to be splitted also as they will differ also between typescript version once the definitions itself are version specific.

It would be also good to have a better separation between the subfolders created for typescript versions and that ones created because of node versions.

**Review by SimonSchick — APPROVED** [review]
I tried this a few months ago but ran into a few roadblocks (tests weren't able to import modules for some reason), good job!

This will make maintaining these typings drastically easier!

Some notes:
We should probably try to avoid defining modules in `global` and instead try to have `global` import those from the respective module definitions, see module, process, etc.
Optional: We can probably cleanup imports a fair bit, eg. replace wildcard imports with more specific ones.

**Comment by rbuckton** [comment]
Unfortunately, `"typesVersions"` is the only way that we can fix the node definitions to allow us to make changes to core lib definitions in the compiler that would conflict with the forward definitions in the node package. Even if we move the minimum TypeScript version forward to 2.8 we would still have this issue as some of the core lib definition changes we would like to make require features not present until at least 3.0.

We can discuss this more within the TypeSCript team, but the folder structure we are using now is what we've decided on and are currently enforcing in both dtslint and types-publisher.

**Comment by SimonSchick** [comment]
@Flarna is 2.x even officially supported anymore eg. is 2.x on LTS?
I cannot find any information on this.

**Comment by SimonSchick** [comment]
Also once this PR lands I will look into updating typings for node v11.

**Comment by Flarna** [comment]
I did a closer look into the changes and tried them in my local setup - it works.

I would also prefer to move at least parts from global.d.ts (e.g. `NodeJS.process`, `NodeJS.ConsoleConstructor`, `NodeJS.Module` into the corresponding module files and keep in globals only the real globals. If there is anything left a common.d.ts could be used but not sure if this is needed.

**Comment by rbuckton** [comment]
@Flarna: Should it matter that the definition for `"process"` in node is literally `module.exports = process`?

**Comment by rbuckton** [comment]
> Also, should test files be split too?

If necessary we should do this in a follow-on PR.

**Review by SimonSchick — APPROVED** [review]

**Comment by rbuckton** [comment]
Tests look good, the only failing tests are due to the recent addition of `readonly` for arrays/tuples that is in the latest dev builds.

**Comment by Flarna** [comment]
Honestly speaking I don't know why the definition of node (e.g. use namespace NodeJS,..) is done that way. I guess because it was created before typescript 2.0 even existed and at that time some augmentations of e.g. classes or so were not possible.
As there are now a lot modules relying on this namespace it's anyway mostly impossible to improve/change this.
What could be done is to move e.g. process specific parts of namespace NodeJS to the process.d.ts file.

**Comment by rbuckton** [comment]
@Flarna, @SimonSchick I may have to revert this change and go back to the previous approach to support `"typesVersions"`, as this change seems to be causing problems for some users (see #32720).

**Comment by SimenB** [comment]
>As there are now a lot modules relying on this namespace it's anyway mostly impossible to improve/change this.

Right now it's also not really possible for libraries to move away from it either (from my understanding). E.g.

```ts
import {ReadStream} from 'tty';
import process from 'process';

const stream: ReadStream = process.stdin;
```

gets `error TS2740: Type 'ReadStream' is missing the following properties from type 'ReadStream': connect, setTimeout, setNoDelay, setKeepAlive, and 17 more.`.

**Comment by SimonSchick** [comment]
I think the biggest painpoint/reason that this namespace exists was that global definitions cannot use imports from modules?

But yea, getting rid of the `NodeJS` namespace would be a big deal, we would have to wait for nodev12 minimum.

**Comment by almeidathomas92** [comment]
> I did a closer look into the changes and tried them in my local setup - it works.
> 
> 
> 
> I would also prefer to move at least parts from global.d.ts (e.g. `NodeJS.process`, `NodeJS.ConsoleConstructor`, `NodeJS.Module` into the corresponding module files and keep in globals only the real globals. If there is anything left a common.d.ts could be used but not sure if this is needed.
> 
>

**Comment by almeidathomas92** [comment]
> @Flarna: Should it matter that the definition for `"process"` in node is literally `module.exports = process`?

**Comment by almeidathomas92** [comment]
> > Also, should test files be split too?
> 
> 
> 
> If necessary we should do this in a follow-on PR.

**Pull request merged**